[MRG]  Fallback for safe_sparse_dot w/ dense output larger than 2**31 elements

rth opened the pull request:
This partially fixes #8666 by providing a fallback `csr_matrix.toarray()` implementation when  the `safe_sparse_dot(.., dense_output=True)` would return an array with more than 2**31 elements. Only CSR arrays are supported.

Performance wise this fallback implementation is between 35% faster (for large sparse arrays) to 40 % slower (for dense arrays converted to the CSR sparse format) than the reference `toarray()` method.
```py
In [1]: import numpy as np
   ...: import scipy.sparse
   ...:
   ...: X = scipy.sparse.rand(5000, 100000, format='csr')
   ...:
   ...: %timeit X.toarray()
   ...: %timeit _csr_to_array_fallback(X)
1 loop, best of 3: 1.88 s per loop
1 loop, best of 3: 1.24 s per loop

In [2]: X = scipy.sparse.csr_matrix(np.random.rand(1000, 5000))
   ...:
   ...: %timeit X.toarray()
   ...: %timeit _csr_to_array_fallback(X)
   ...:
10 loops, best of 3: 33.5 ms per loop
10 loops, best of 3: 45.6 ms per loop
```

**PS:** codecov is unhappy, again.. sigh 
-0.01% patch coverage: I'm sure it deserves a failed CI status when you look at the list of PRs...

rth committed:
Add fallback for csr_matrix.toarray()

rth committed:
Fix random state issue for sparse arrays with old scipy versions

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -174,6 +174,24 @@ def density(w, **kwargs):
+        row_indices[indptr[idx]:indptr[idx+1]] = idx
+    Y = np.zeros(X.shape, dtype=X.dtype)
+    Y[row_indices, X.indices] = X.data
+    return Y
```
I would rather put that function into `sklearn/utils/fixes.py` which is dedicated to numpy and scipy backports.

ogrisel commented on sklearn/utils/tests/test_extmath.py:
```diff
@@ -661,3 +663,20 @@ def test_stable_cumsum():
+
+
+def test_csr_to_array_fallback():
+    X = np.random.rand(100, 10)
```
Please make test run deterministically and independently of one another by default:

```python
X = np.random.RandomState(42).rand(100, 10)
```

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -174,6 +174,24 @@ def density(w, **kwargs):
+    elements.
+    """
+    if X.format != 'csr':
+        raise ValueError('Only CSR sparse arrays are supported')
```
This is a private method with an explicit name and the only location where we use it in the code does the check on `X.format` explicitly. Let's get rid of this untested input check.

If we would like to keep the input check, it should be more informative (put the actual value of `X.format` in the error message) and should be tested.

ogrisel reviewed (commented):
I had a quick look (see comments). I am not sure about this. This is a bug of scipy so a backport seems legit. However I am pretty sure that any occurrence of that bug are likely to be misuse of scikit-learn estimators that would raise a MemoryError anyway.

Arguably a `MemoryError` is easier to understand that the current cryptic message so this PR is good in that respect. +0 for merge on my side once the comments are addressed.

rth committed:
Addressed review comments

rth commented:
@ogrisel Thanks for the review! I addressed your comments.

> I am not sure about this. This is a bug of scipy so a backport seems legit. However I am pretty sure that any occurrence of that bug are likely to be misuse of scikit-learn estimators that would raise a MemoryError anyway.

Well, essentially this would mostly affect sparse dot product calculations, if one takes two sparse arrays`X, Y` of shape  `(n_samples, n_ features)` and compute the dot product `X.dot(Y.T)`, the sparsity of the output decreases with `n_features`,
```py
import numpy as np
import scipy.sparse

n0 = 1000
for n in [1000, 5000, 10000, 20000]:
    np.random.seed(n)
    X = scipy.sparse.rand(n0, n, format='csr')
    Y = scipy.sparse.rand(n0, n, format='csr')
    print(f'Array size: {X.shape}')
    print(f'Sparsity: input: {X.nnz/(n0*n)} {Y.nnz/(n0*n)} => ', end='')
    Z = X.dot(Y.T)
    print(f'output: {Z.nnz/(n0*n0)}')
```
```
Array size: (1000, 1000)
Sparsity: input: 0.01 0.01 => output: 0.095556
Array size: (1000, 5000)
Sparsity: input: 0.01 0.01 => output: 0.393677
Array size: (1000, 10000)
Sparsity: input: 0.01 0.01 => output: 0.632454
Array size: (1000, 20000)
Sparsity: input: 0.01 0.01 => output: 0.864754
```
so for `n_features` large enough, the end of the `safe_sparse_dot` one might end up with a CSR array which is almost dense.  And then reaching the 2**31 elements limit when calling `safe_sparse_dot(X, Y.T, dense_output=True)` is not that difficult, say `X.shape[0] == 2.5e6` , `Y.shape[0] == 1000` .
In other cases when calling `csr_matrix.to_array()` this is mostly without incidence, I think.

But since no-one reported this issue up to now, and since this would be  fixed in the upcoming scipy version, I agree that a backport might be worth it. So I'm closing this PR for now. Thanks for the feedback!

rth commented:
@ogrisel OK, I get your points that one shouldn't have to compute pairwise distances in sparse high dimensional space anyway because of the curse of dimentionality. 

But `safe_sparse_dot(.. , dense_output=True)` is also used elsewhere where this issue could potentially arise,
  -  [Rige regression with the Cholesky solver]( https://github.com/scikit-learn/scikit-learn/blob/e3c9ae204ffb152c151e9b61306ff8f16a2c1e0a/sklearn/linear_model/ridge.py#L395)
  -  SGDRegressor and ElasticNet [decision_function](https://github.com/scikit-learn/scikit-learn/blob/e3c9ae204ffb152c151e9b61306ff8f16a2c1e0a/sklearn/linear_model/stochastic_gradient.py#L991) though one would need indeed to have  a sparse test set with `n_samples=>2.3e9` to see it (I guess it's not very common)...
